transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java

```diff
@@ -157,12 +158,71 @@ public final void read() {
         }
     }
 
+    private enum WriteStatus {
+        SET_OP_WRITE, // We have not written out the whole message and output channel is not writable
+        INCOMPLETE, // We have not written out the whole message after the spin count
+        DONE // we have written out the whole message
+    }
+
+    // returns whether we should setOpWrite
+    private WriteStatus doWriteFileRegion(ChannelOutboundBuffer in, FileRegion region,
+            int writeSpinCount) throws Exception {
+        if (region instanceof SliceableFileRegion) {
+            SliceableFileRegion sliceableRegion = (SliceableFileRegion) region;
+            if (!sliceableRegion.isTransferable()) {
+                return WriteStatus.DONE;
+            }
+            WriteStatus status = WriteStatus.INCOMPLETE;
+            long flushedAmount = 0;
+
+            for (int i = writeSpinCount - 1; i >= 0; i--) {
+                long localFlushedAmount = doWriteFileRegion(sliceableRegion);
+                if (localFlushedAmount == 0) {
+                    status = WriteStatus.SET_OP_WRITE;
+                    break;
+                }
+
+                flushedAmount += localFlushedAmount;
+                if (!sliceableRegion.isTransferable()) {
+                    status = WriteStatus.DONE;
+                    break;
+                }
+            }
+            in.progress(flushedAmount);
+            return status;
+        } else {
+            // TODO: consider remove this branch once we move the methods in SliceableFileRegion
+            // into FileRegion in the next minor release.
+            if (region.transferred() >= region.count()) {
+                return WriteStatus.DONE;
+            }
+            WriteStatus status = WriteStatus.INCOMPLETE;
+            long flushedAmount = 0;
+
+            for (int i = writeSpinCount - 1; i >= 0; i--) {
+                long localFlushedAmount = doWriteFileRegion(region);
+                if (localFlushedAmount == 0) {
+                    status = WriteStatus.SET_OP_WRITE;
+                    break;
+                }
+
+                flushedAmount += localFlushedAmount;
+                if (region.transferred() >= region.count()) {
+                    status = WriteStatus.DONE;
+                    break;
+                }
+            }
+            in.progress(flushedAmount);
+            return status;
+        }
+    }
+
     @Override
     protected void doWrite(ChannelOutboundBuffer in) throws Exception {
         int writeSpinCount = -1;
 
         boolean setOpWrite = false;
-        for (;;) {
+        LOOP: for (;;) {
```

Comment:
Upper case should be reserved for constants.  is a label really a constant?

Reply:
This is only my style...

What do we suggest in netty? 'Loop'? 'loop'?

Reply:
`loop:`

Reply:
Done.